extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -1715,9 +1726,109 @@ private void publishSegmentsIfNeeded(
       //noinspection unchecked
       @SuppressWarnings("unchecked")
       final Set<DataSegment> segments = (Set<DataSegment>) queryKernel.getResultObjectForStage(finalStageId);
+
+      Function<Set<DataSegment>, Set<DataSegment>> compactionStateAnnotateFunction = Function.identity();
```

Comment:
Rather than keeping this outside the if-else and assigning `identity()` to it. It should just be used where necessary. You can make the `segments` non-final.

Reply:
Applying the function in the branch itself now before sending to publish.